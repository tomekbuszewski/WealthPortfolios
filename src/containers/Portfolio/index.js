import * as React from 'react';
import Props from 'prop-types';

import Communication from '@src/services/Communication';

import Loader from '@src/components/Loader';
import View from '@src/views/Portfolio';

import { ENDPOINTS } from '@config';

class Portfolio extends React.Component {
  constructor(props) {
    super(props);

    const { match: { params: { id } } } = this.props;
    this.url = `${ENDPOINTS.PORTFOLIOS}/${id}`;

    this.fetch = new Communication({
      url: this.url,
    });

    this.state = {
      data: undefined,
      editable: false,
      success: null,
    };
  }

  static get propTypes() {
    return {
      match: Props.shape({
        params: Props.shape({
          id: Props.string,
        }),
      }).isRequired,
    };
  }

  componentDidMount() {
    this.fetch().then(({ data }) => this.setState({ data }));
  }

  changeData = (event) => {
    const { name, value } = event.target;
    const { data } = this.state;

    this.setState({
      data: {
        ...data,
        [name]: value,
      },
    });
  }

  toggleEdit = () => {
    const { editable } = this.state;

    this.setState({ editable: !editable }, () => {
      const { editable: newEditable } = this.state;

      if (newEditable === false) {
        this.sendModifications();
      }
    });
  }

  resetStatus = () => {
    setTimeout(() => {
      this.setState({ success: null });
    }, 2000);
  }

  sendModifications() {
    const { data } = this.state;

    const put = new Communication({
      url: this.url,
      data,
      method: 'put',
    });

    put()
      .then(result => this.setState({ success: result.status === 200 }, this.resetStatus))
      .catch(() => this.setState({ success: false }, this.resetStatus));
  }

  render() {
    const { data, editable, success } = this.state;

    return typeof data !== 'undefined'
      ? (
        <View
          data={data}
          editable={editable}
          editToggle={this.toggleEdit}
          onChange={this.changeData}
          success={success}
        />
      )
      : <Loader />;
  }
}

export default Portfolio;
