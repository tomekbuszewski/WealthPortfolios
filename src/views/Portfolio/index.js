import * as React from 'react';
import Props from 'prop-types';

import Input from '@src/components/Input';
import Section from '@src/components/Section';
import Notify from '@src/components/Notification';

const Portfolio = (props) => {
  const {
    data,
    editable,
    editToggle,
    onChange,
    success,
  } = props;

  return (
    <Section title={`${data.name} â€” ${data.clientName}`}>
      <article>
        {success === true && <Notify>Edit successful</Notify>}
        {success === false && <Notify type="warning">Edit failed</Notify>}
        <button className="button is-small" type="button" onClick={editToggle}>{editable ? 'Finish' : 'Edit'}</button>
        <form>
          <Input type="number" value={data.value} name="value" label="Amount" disabled={!editable} onChange={onChange} />
          <Input type="text" value={data.currency} name="currency" label="Currency" disabled={!editable} onChange={onChange} />
          <Input type="text" value={data.name} name="name" label="Name" disabled={!editable} onChange={onChange} />
          <Input type="text" value={data.clientName} name="clientName" label="Client" disabled={!editable} onChange={onChange} />
        </form>
      </article>
    </Section>
  );
};

Portfolio.propTypes = {
  data: Props.shape({
    clientName: Props.string,
    currency: Props.string,
    id: Props.oneOfType([Props.string, Props.number]),
    name: Props.string,
    value: Props.oneOfType([Props.string, Props.number]),
  }).isRequired,
  editable: Props.bool.isRequired,
  editToggle: Props.func.isRequired,
  success: Props.bool,
  onChange: Props.func.isRequired,
};

export default Portfolio;
