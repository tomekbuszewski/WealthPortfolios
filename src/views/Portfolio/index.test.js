import * as React from 'react';
import renderer from 'react-test-renderer';

import Portfolio from './';

import data from '@mocks/portfolios';

describe ('Portfolio tests', () => {
  it ('should match snapshot with editable false', () => {
    const component = renderer.create(<Portfolio
      editable={false}
      editToggle={() => false}
      data={data[0]}
      onChange={() => false}
    />);
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it ('should match snapshot with editable true', () => {
    const component = renderer.create(<Portfolio
      editable={true}
      editToggle={() => false}
      data={data[0]}
      onChange={() => false}
    />);
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
