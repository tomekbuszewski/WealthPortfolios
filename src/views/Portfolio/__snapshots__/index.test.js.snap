// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portfolio tests should match snapshot with editable false 1`] = `
<section
  className="sc-bdVaJa gPOcxX section hero"
>
  <header
    className="title"
  >
    Portfolio 1 — Kowalski
  </header>
  <div
    className="content is-large"
  >
    <article>
      <button
        className="button is-small"
        onClick={[Function]}
        type="button"
      >
        Edit
      </button>
      <form>
        <label
          className="label"
          htmlFor="value"
        >
          Amount
          <input
            className="input"
            disabled={true}
            id="value"
            name="value"
            onChange={[Function]}
            type="text"
            value={100}
          />
        </label>
        <label
          className="label"
          htmlFor="currency"
        >
          Currency
          <input
            className="input"
            disabled={true}
            id="currency"
            name="currency"
            onChange={[Function]}
            type="text"
            value="PLN"
          />
        </label>
        <label
          className="label"
          htmlFor="name"
        >
          Name
          <input
            className="input"
            disabled={true}
            id="name"
            name="name"
            onChange={[Function]}
            type="text"
            value="Portfolio 1"
          />
        </label>
        <label
          className="label"
          htmlFor="clientName"
        >
          Client
          <input
            className="input"
            disabled={true}
            id="clientName"
            name="clientName"
            onChange={[Function]}
            type="text"
            value="Kowalski"
          />
        </label>
      </form>
    </article>
  </div>
</section>
`;

exports[`Portfolio tests should match snapshot with editable true 1`] = `
<section
  className="sc-bdVaJa gPOcxX section hero"
>
  <header
    className="title"
  >
    Portfolio 1 — Kowalski
  </header>
  <div
    className="content is-large"
  >
    <article>
      <button
        className="button is-small"
        onClick={[Function]}
        type="button"
      >
        Finish
      </button>
      <form>
        <label
          className="label"
          htmlFor="value"
        >
          Amount
          <input
            className="input"
            disabled={false}
            id="value"
            name="value"
            onChange={[Function]}
            type="text"
            value={100}
          />
        </label>
        <label
          className="label"
          htmlFor="currency"
        >
          Currency
          <input
            className="input"
            disabled={false}
            id="currency"
            name="currency"
            onChange={[Function]}
            type="text"
            value="PLN"
          />
        </label>
        <label
          className="label"
          htmlFor="name"
        >
          Name
          <input
            className="input"
            disabled={false}
            id="name"
            name="name"
            onChange={[Function]}
            type="text"
            value="Portfolio 1"
          />
        </label>
        <label
          className="label"
          htmlFor="clientName"
        >
          Client
          <input
            className="input"
            disabled={false}
            id="clientName"
            name="clientName"
            onChange={[Function]}
            type="text"
            value="Kowalski"
          />
        </label>
      </form>
    </article>
  </div>
</section>
`;
